core/store/src/trie/mem/resharding.rs

```diff
@@ -0,0 +1,250 @@
+    /// retain operation.
+    fn retain_multi_range(mut self, intervals: &[Range<Vec<u8>>]) -> (TrieChanges, TrieAccesses) {
+        debug_assert!(intervals.iter().all(|range| range.start < range.end));
+        let intervals_nibbles = intervals
+            .iter()
+            .map(|range| {
+                NibbleSlice::new(&range.start).iter().collect_vec()
+                    ..NibbleSlice::new(&range.end).iter().collect_vec()
+            })
+            .collect_vec();
+        // let root = self.get_root().unwrap();
+        // TODO(#12074): consider handling the case when no changes are made.
+        // TODO(#12074): restore proof as well.
+        self.retain_multi_range_recursive(0, vec![], &intervals_nibbles);
+        self.to_trie_changes()
+    }
+
+    /// Recursive implementation of the algorithm of retaining keys belonging to
+    /// any of the ranges given in `intervals` from the trie.
+    ///
+    /// `node_id` is the root of subtree being explored.
+    /// `key_nibbles` is the key corresponding to `root`.
+    /// `intervals_nibbles` is the list of ranges to be retained.
+    ///
+    /// Returns id of the node after retain applied.
+    fn retain_multi_range_recursive(
+        &mut self,
+        node_id: usize,
+        key_nibbles: Vec<u8>,
+        intervals_nibbles: &[Range<Vec<u8>>],
+    ) {
+        let decision = retain_decision(&key_nibbles, intervals_nibbles);
+        match decision {
+            RetainDecision::RetainAll => return,
+            RetainDecision::DiscardAll => {
+                let _ = self.take_node(node_id);
+                self.place_node(node_id, UpdatedMemTrieNode::Empty);
+                return;
+            }
+            RetainDecision::Descend => {
+                // We need to descend into all children. The logic follows below.
+            }
+        }
+
+        let node = self.take_node(node_id);
+        match node {
+            UpdatedMemTrieNode::Empty => {
+                // Nowhere to descend.
+                self.place_node(node_id, UpdatedMemTrieNode::Empty);
+                return;
+            }
+            UpdatedMemTrieNode::Leaf { extension, value } => {
+                let full_key_nibbles =
+                    [key_nibbles, NibbleSlice::from_encoded(&extension).0.iter().collect_vec()]
+                        .concat();
+                if !intervals_nibbles.iter().any(|interval| interval.contains(&full_key_nibbles)) {
+                    self.place_node(node_id, UpdatedMemTrieNode::Empty);
+                } else {
+                    self.place_node(node_id, UpdatedMemTrieNode::Leaf { extension, value });
+                }
+            }
+            UpdatedMemTrieNode::Branch { mut children, mut value } => {
+                if !intervals_nibbles.iter().any(|interval| interval.contains(&key_nibbles)) {
+                    value = None;
+                }
+
+                for i in 0..16 {
```

Comment:
nit: could we do `for i in 0..children.len()` or `for (i, child) in mut children.iter_mut().enumerate()`?